aderyn_core/src/detect/high/uninitialized_state_variables.rs

```diff
@@ -0,0 +1,156 @@
+use std::collections::{BTreeMap, HashSet};
+use std::error::Error;
+
+use crate::{
+    ast::{Expression, FunctionDefinition, Statement},
+    context::loader::{ASTNode, ContextLoader},
+    detect::detector::{Detector, IssueSeverity},
+};
+use eyre::Result;
+
+#[derive(Default)]
+pub struct UninitializedStateVariablesDetector {
+    // Keys are source file name and line number
+    found_instances: BTreeMap<(String, usize), String>,
+}
+
+impl UninitializedStateVariablesDetector {
+    fn is_state_variable(&mut self, expression: &Expression) -> Option<String> {
+        // Implement logic to check if the expression is a state variable
+        // For instance, if `expression` is an Identifier or MemberAccess,
+        // determine if it refers to a state variable.
+        // Return Some(variable_name) if it's a state variable, None otherwise.
+        match expression {
+            Expression::Identifier(identifier) => Some(identifier.name.clone()),
+            Expression::MemberAccess(member_access) => {
+                if member_access.referenced_declaration.is_some() {
+                    Some(member_access.member_name.clone())
+                } else {
+                    None
+                }
+            }
+            _ => None,
+        }
+    }
+
+    fn function_uses_variable_without_initialization(
+        &mut self,
+        function_node: &FunctionDefinition,
+        variable_name: &str,
+    ) -> bool {
+        // Implement logic to check if the function uses the given variable name without it being initialized
+        // This would typically involve parsing the function body and looking for usage of the variable
+        // Return true if the variable is used without being initialized, false otherwise
+        let function_body = function_node.body.as_ref();
+        for statement in &function_body.unwrap().statements {
+            match statement {
+                Statement::ExpressionStatement(expression_statement) => {
+                    let expression = expression_statement.expression.to_owned();
+                    match expression {
+                        Expression::MemberAccess(member_access) => {
+                            if member_access.member_name == variable_name {
+                                return true;
+                            }
+                        }
+                        _ => {}
+                    }
+                }
+                _ => {}
+            }
+        }
+
+        // If no uninitialized usage is found, return false.
+        false
+    }
+}
+
+impl Detector for UninitializedStateVariablesDetector {
+    fn detect(&mut self, loader: &ContextLoader) -> Result<bool, Box<dyn Error>> {
```

Comment:
Good start, but there's probably an easier way to approach this one. Instead of starting with the assignments, what if it:

1. Get all state variables that are not hardcoded with a value - into a variable called `found_state_variables`
2. Get all Assignments - `found_assignments`
3. For each of the Assignments, if the left-hand side references a `found_state_variable`, remove it from the `found_state_variables`.

Whatever is left inside `found_state_variables, add them to the `instances`. W probably don't need the extra `impl` above, and we can add specifically the state variables declaration to the `instances`.

Make sense?

Reply:
So I decided to check all function bodies independently as an assignment in one would mean that uninitialized usage in another would not be considered a found instance. Are you suggesting that perhaps this is desired, e.g. admin setter function exists, therefore usage in other functions should not be uninitialized and we should not detect this as an issue?

Reply:
Yep, once we have that working, we can discuss what it finds and whether or not we should rethink how granular it should be